annotated.go

```diff
@@ -89,3 +90,228 @@ func (a Annotated) String() string {
+	annotatedResult := []reflect.StructField{{
+		Name:      "Out",
+		Type:      reflect.TypeOf(Out{}),
+		Anonymous: true,
+	}}
+
+	for i := 0; i < fType.NumOut(); i++ {
+		structField := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", i),
+			Type: fType.Out(i),
+		}
+		annotatedResult = append(annotatedResult, structField)
+	}
+
+	resultTypes := []reflect.Type{reflect.StructOf(annotatedResult)}
+	return resultTypes
+}
+
+// ResultTags is an Annotation that annotates the result(s) of a function.
+// When multiple tags are specified, each tag is mapped to the corresponding
+// positional result.
+func ResultTags(tags ...string) Annotation {
+	r := resultTags{tags}
+	return r
+}
+
+// Annotate lets you annotate a function's paramter and returns with tags
+// without you having to declare separate struct definitions for them.
+//
+// For example,
+//   func NewGateway(ro, rw *db.Conn) *Gateway { ... }
+//   fx.Provide(
+//     fx.Annotate(
+//       NewGateway,
+//       fx.ParamTags(`name:"ro" optional:"true"`, `name:"rw"`),
+//       fx.ResultTags(`name:"foo"`),
+//     ),
+//   )
+//
+// Is equivalent to,
+//
+//  type params struct {
+//    fx.In
+//
+//    RO *db.Conn `name:"ro" optional:"true"`
+//    RW *db.Conn `name:"rw"`
+//  }
+//
+//  type result struct {
+//    fx.Out
+//
+//    GW *Gateway `name:"foo"`
+//   }
+//
+//   fx.Provide(func(p params) result {
+//     return result{GW: NewGateway(p.RO, p.RW)}
+//   })
+//
+// If there are more Annotations are specified than than one per
+// Annotation type, none of the Annotations will be applied.
+//
+// For example,
+//
+// fx.Provide(
+//   fx.Annotate(
+//     NewGateWay,
+//     fx.ParamTags(`name:"ro" optional:"true"`),
+//     fx.ParamTags(`name:"rw"),
+//     fx.ResultTags(`name:"foo"`)
+//   )
+// )
+//
+// is considered an invalid usage and will not apply any of the
+// Annotations to NewGateway.
+//
+// If more tags are given than the number of parameters/results, only
+// the ones up to the number of parameters/results will be applied.
+func Annotate(f interface{}, anns ...Annotation) interface{} {
+	fVal := reflect.ValueOf(f)
+	fType := fVal.Type()
+	numIn := fType.NumIn()
+	numOut := fType.NumOut()
+
+	if !verifyAnnotation(numIn, numOut, anns...) {
+		return f
+	}
+
+	var ins []reflect.Type
+	var outs []reflect.Type
+
+	annotatedIn := false
+	annotatedOut := false
+
+	for _, ann := range anns {
+		if pTags, ok := ann.(paramTags); ok {
+			ins = pTags.getAnnotatedType(fType)
+			annotatedIn = true
+		}
+		if rTags, ok := ann.(resultTags); ok {
+			outs = rTags.getAnnotatedType(fType)
+			annotatedOut = true
+		}
+	}
+
+	if !annotatedIn {
+		ins = make([]reflect.Type, numIn)
+		for i := 0; i < fType.NumIn(); i++ {
+			ins[i] = fType.In(i)
+		}
+	}
+	if !annotatedOut {
+		outs = make([]reflect.Type, numOut)
+		for i := 0; i < fType.NumOut(); i++ {
+			outs[i] = fType.Out(i)
+		}
+	}
+
+	var newF func([]reflect.Value) []reflect.Value
+
+	if annotatedIn && annotatedOut {
```

Comment:
I suspect we can clean this up a bit. If we annotated the input, then we need a `func([]reflect.Value) []reflect.Value` that turns the fake fx.In struct into a list of parameters. If we annotated the output, we need one that turns a list of results into an fx.Out struct. The default implementation is "return as-is" for when we didn't annotate. Using that, we'll have only one code block generating `newF`.